File: server/pkg/biz/assert.go
package biz

import (
	"context"
	"mayfly-go/pkg/errorx"
	"mayfly-go/pkg/i18n"
	"mayfly-go/pkg/utils/anyx"

	"reflect"
)

// 断言错误为ni
// @param msgAndParams 消息与参数占位符，第一位为错误消息可包含%s等格式化标识。其余为Sprintf格式化值内容
//
//	ErrIsNil(err)
//	ErrIsNil(err, "xxxx")
//	ErrIsNil(err, "xxxx: %s", "yyyy")
func ErrIsNil(err error, msgAndParams ...any) {
	if err != nil {
		if len(msgAndParams) == 0 {
			panic(errorx.NewBiz(err.Error()))
		}

		panic(errorx.NewBizf(msgAndParams[0].(string), msgAndParams[1:]...))
	}
}

// 断言错误为ni
// @param msgId i18n消息id
//
//	ErrIsNil(err)
//	ErrIsNil(err, "xxxx")
//	ErrIsNil(err, "xxxx: %s", "yyyy")
func ErrIsNilI(ctx context.Context, err error, msgId i18n.MsgId, attrs ...any) {
	if err != nil {
		if len(attrs) == 0 {
			panic(errorx.NewBiz(err.Error()))
		}

		panic(errorx.NewBiz(i18n.TC(ctx, msgId, attrs...)))
	}
}

func ErrNotNil(err error, msg string, params ...any) {
	if err == nil {
		panic(errorx.NewBizf(msg, params...))
	}
}

// ErrIsNilAppendErr 断言错误为nil，否则抛出业务异常，异常消息可包含‘%s’进行错误信息提示
//
//	// -> xxxx: err.Error()
//	biz.ErrIsNilAppendErr(err, "xxxx: %s")
func ErrIsNilAppendErr(err error, msg string) {
	if err != nil {
		panic(errorx.NewBizf(msg, err.Error()))
	}
}

// ErrIsNilAppendErr 断言错误为nil，否则抛出业务异常，异常消息可包含‘%s’进行错误信息提示
//
//	// -> xxxx: err.Error()
//	biz.ErrIsNilAppendErr(err, "xxxx: %s")
func ErrIsNilAppendErrI(ctx context.Context, err error, msgId i18n.MsgId) {
	if err != nil {
		panic(errorx.NewBizf(i18n.TC(ctx, msgId), err.Error()))
	}
}

func IsTrue(exp bool, msg string, params ...any) {
	if !exp {
		panic(errorx.NewBizf(msg, params...))
	}
}

func IsTrueI(ctx context.Context, exp bool, msgId i18n.MsgId, attrs ...any) {
	if !exp {
		panic(errorx.NewBiz(i18n.TC(ctx, msgId, attrs...)))
	}
}

func IsTrueBy(exp bool, err *errorx.BizError) {
	if !exp {
		panic(err)
	}
}

func NotEmpty(str string, msg string, params ...any) {
	if str == "" {
		panic(errorx.NewBizf(msg, params...))
	}
}

func NotNil(data any, msg string, params ...any) {
	if reflect.ValueOf(data).IsNil() {
		panic(errorx.NewBizf(msg, params...))
	}
}

func NotBlank(data any, msg string, params ...any) {
	if anyx.IsBlank(data) {
		panic(errorx.NewBizf(msg, params...))
	}
}

func IsEquals(data any, data1 any, msg string) {
	if data != data1 {
		panic(errorx.NewBiz(msg))
	}
}

func Nil(data any, msg string) {
	if !reflect.ValueOf(data).IsNil() {
		panic(errorx.NewBiz(msg))
	}
}

File: mayfly_go_web/src/common/assert.ts
/**
 * 不符合业务断言错误
 */
class AssertError extends Error {
    constructor(message: string) {
        super(message);
        // 错误类名
        this.name = 'AssertError';
    }
}

/**
 * 断言表达式为true
 *
 * @param condition 条件表达式
 * @param msg 错误消息
 */
export function isTrue(condition: boolean, msg: string) {
    if (!condition) {
        throw new AssertError(msg);
    }
}

/**
 * 断言不能为空值，即null,0,''等
 *
 * @param obj 对象1
 * @param msg 错误消息
 */
export function notBlank(obj: any, msg: string) {
    isTrue(obj, msg);
}

/**
 * 断言两对象相等
 *
 * @param obj1 对象1
 * @param obj2 对象2
 * @param msg 错误消息
 */
export function isEquals(obj1: any, obj2: any, msg: string) {
    isTrue(obj1 === obj2, msg);
}

/**
 * 断言对象不为null或undefiend
 *
 * @param obj 对象
 * @param msg 错误提示
 */
export function notNull(obj: any, msg: string) {
    if (obj == null || obj == undefined) {
        throw new AssertError(msg);
    }
}

/**
 * 断言字符串不能为空
 *
 * @param str 字符串
 * @param msg 错误提示
 */
export function notEmpty(str: string, msg: string) {
    if (str == null || str == undefined || str == '') {
        throw new AssertError(msg);
    }
}
